include/npcomp/E2E/Passes.td

```diff
@@ -83,4 +78,33 @@ def LowerToLLVM : Pass<"e2e-lower-to-llvm", "ModuleOp"> {
   let constructor = "mlir::NPCOMP::createLowerToLLVMPass();";
 }
 
+// TODO: Move this pass to upstream.
+// TODO: This pass will still do "folding" on all ops.
+// The applyPatternsAndFoldGreedily driver will need to be changed to restrict
+// folding to the specified dialects as well.
+// Perhaps a better design is having a pass that uses the conversion framework.
+// The the pass constructor would take a set of op names, and it would
+// set up a conversion target that makes all those ops illegal, and uses
+// the canonicalization patterns from those ops to legalize them.
+def RestrictedCanonicalizer : Pass<"restricted-canonicalize"> {
+  let summary = "Canonicalize operations";
+  let description = [{
+    This pass is the same as the regular `canonicalize` pass, but it only
+    applies a restricted set of patterns.
+
+    This is useful when a particular canonicalization is actually needed for
```

Comment:
I've needed something like this enough times... might be worth seeing how to kick it upstream.

Reply:
Will do.